### ISSUE TITLE
BUG: crs is lost after calling GeoDataFrame.apply() method

### ISSUE THREAD

**Issue by galak75** [geopandas/geopandas#902]
crs is null on `GeoDataFrame.apply(...)` result.
the following unit test fails because `result.crs` is `None`:

```
    def test_apply_function_to_rows():
        gdf = GeoDataFrame(
            data={'foo': ['Bar']},
            geometry=[Point(45.5089192, -73.5565212)],
            crs={'init': 'epsg:4326'}
        )

        def _lower_foo(row):
            row['foo'] = row['foo'].lower()
            return row

        result = gdf.apply(_lower_foo, axis=1, reduce=False)

        assert_geodataframe_equal(
            result,
            GeoDataFrame(
                data={'foo': ['bar']},
                geometry=[Point(45.5089192, -73.5565212)],
                crs={'init': 'epsg:4326'}
            )
        )
```
running with the environment below:
```
$ pip list
Package         Version    
--------------- -----------
atomicwrites    1.2.1      
attrs           18.2.0     
Click           7.0        
click-plugins   1.0.4      
cligj           0.5.0      
Fiona           1.8.4      
geopandas       0.4.0      
more-itertools  5.0.0      
munch           2.3.2      
numpy           1.16.0     
pandas          0.23.4     
pip             19.0.1     
pluggy          0.8.1      
py              1.7.0      
pyproj          1.9.6      
pytest          4.1.1      
python-dateutil 2.7.5      
pytz            2018.9     
setuptools      40.6.3     
Shapely         1.6.4.post2
six             1.12.0     
```

**Comment by jorisvandenbossche** [comment]
The problem with trying to preserve the crs, in general, is that `apply` is a very generic method for which we cannot make much assumptions about what the result will be.

In your example, you keep the geometry column, but often the result of an apply call will not even have the geometry anymore, or might have transformed the geometry.

So I am not fully sure we should try to preserve it.

**Comment by spolloni** [comment]
just unexpectedly encountered this behavior. I'm inclined to think users shouldn't invariably have to re-set the crs after using `apply()`.  Seems fair to assume on average only a small fraction of arguments are geometry-altering functions, and perhaps the burden of caution should be placed on those uses. I do a agree it's a contentious issue though...

**Comment by martinfleis** [comment]
ref #1351

**Comment by JessicaS11** [comment]
Like @spolloni, I just unexpectedly encountered this behavior. I was surprised to discover that `gdf.apply()` actually resulted in a pandas series being passed to the function rather than an object of type geopandas geoseries. In my case I was trying to work specifically with a named geometry column, so I could not generalize my function to access `.geometry`. It was easy enough to work around, but I'm now wondering that there's not a geopandas wrapper for a common iterator like `apply()` to pass the geoseries through (or perhaps the more generalizable approach would be to have pandas check for a valid extension type in `apply()` and allow the extended type to pass through if it's valid).

**Comment by martinfleis** [comment]
I honestly thought we fixed this in #1478 but that apparently fixes only `GeoSeries.apply` not `GeoDataFrame.apply`. We should do the similar wrapper for GeoDataFrame as well then.

**Comment by jorisvandenbossche** [comment]
The original issue about CRS getting lost is fixed in #1848, so closing this issue.

There is one more aspect that @JessicaS11 mentioned, about what object is being passed to the applied function (when applying it to each column, I assume), whether it is a Series or GeoSeries. 

```
In [1]: df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))

In [2]: def func(col):
   ...:     print(type(col))
   ...:     return col
   ...: 

In [3]: df.apply(func)
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
Out[3]: 
       pop_est      continent                      name iso_a3  gdp_md_est                                           geometry
0       920938        Oceania                      Fiji    FJI      8374.0  MULTIPOLYGON (((180.00000 -16.06713, 180.00000...
1     53950935         Africa                  Tanzania    TZA    150600.0  POLYGON ((33.90371 -0.95000, 34.07262 -1.05982...
2       603253         Africa                 W. Sahara    ESH       906.5  POLYGON ((-8.66559 27.65643, -8.66512 27.58948...
3     35623680  North America                    Canada    CAN   1674000.0  MULTIPOLYGON (((-122.84000 49.00000, -122.9742...
4    326625791  North America  United States of America    USA  18560000.0  MULTIPOLYGON (((-122.84000 49.00000, -120.0000...
..         ...            ...                       ...    ...         ...                                                ...
172    7111024         Europe                    Serbia    SRB    101800.0  POLYGON ((18.82982 45.90887, 18.82984 45.90888...
173     642550         Europe                Montenegro    MNE     10610.0  POLYGON ((20.07070 42.58863, 19.80161 42.50009...
174    1895250         Europe                    Kosovo    -99     18490.0  POLYGON ((20.59025 41.85541, 20.52295 42.21787...
175    1218208  North America       Trinidad and Tobago    TTO     43570.0  POLYGON ((-61.68000 10.76000, -61.10500 10.890...
176   13026129         Africa                  S. Sudan    SSD     20880.0  POLYGON ((30.83385 3.50917, 29.95350 4.17370, ...

[177 rows x 6 columns]
```

So the last column is also passed as a Series and not as GeoSeries. 

I am not sure that is easy to fix (override) on the GeoPandas side, though. Should take a look at how this is implemented in pandas (eg it might not be using `_constructor` properly).

**Comment by jorisvandenbossche** [comment]
Moved that last comment to a new issue -> https://github.com/geopandas/geopandas/issues/1852
